docs/tutorials/PyTorch.py

```diff
@@ -0,0 +1,346 @@
+#
+# If we were on the side of a hill in the dark and we wanted to get down to the bottom of a valley, how would we do it?
+#
+# We wouldn't be able to see all the way to the bottom of the valley, but we could feel which way is down based on where we are standing. We would take steps in the downward direction and we'd know when to stop when the ground felt flat.
+#
+# One other thing we'd have to consider is our step size. If we take very small steps in the direction of the descent, it will take us a longer time than if we take larger steps. However, if we take super long steps, we might completely miss the flat part of the valley, and start ascending the other side of the valley.
+#
+# We can look at the gradient descent from a more mathematical lense by looking at the graph $y = x^2$:
+
+# +
+def y(x_input):
+    y = x_input ** 2
+    return y
+
+
+x = np.linspace(-5, 5, 100)
+plt.plot(x, y(x))
+# -
+
+# We will choose some arbitrary place to start on the left side of the hill.
+
+# +
+x = np.linspace(-5, 5, 100)
+plt.plot(x, y(x))  # plot y = x ** 2
+
+x_start = -4  # x coordinate of starting point
+y_start = 16  # y coordinate of starting point
+
+plt.scatter(x_start, y_start)  # plot starting point
+plt.show()
+# -
+
+# If we plot a tangent line at this point, we see which directions we can go:
+
+# +
+x = np.linspace(-5, 5, 100)
+plt.plot(x, y(x))  # plot y = x ** 2
+
+
+plt.scatter(x_start, y_start)  # plot starting point
+
+slope_start = 2 * x_start  # derivative of y = x ** 2 evaluated at x_start
+plt.plot(x, slope_start * (x - x_start) + y_start)  # tangent line
+
+plt.xlim(xmin=-5, xmax=5)
+plt.ylim(ymin=0, ymax=25)
+plt.show()
+# -
+
+# We see we need to go down to go toward the minimum. We take a very small step in the direction of the steepest downward slope. When we take steps, we find the x coordinate of our new location by this equation:
+#
+# $x_{new} = x_{current} - \nabla y(x_{current}) * (step \: size)$
+#
+# where:
+#
+# - $x_{current}$ is our current x value
+#
+# - $\nabla y(x_{current})$ is the gradient at our current point
+#
+# - $(step \: size)$ is a value we choose that scales our steps
+#
+# This makes sense because we start at some x value and we want to get to a new x value that is closer to the minimum. In the above plot, the gradient is negative. If we were to add the gradient to our current x value that would bring us further away from the minimum to a more negative x value. This is because the gradient points in the direction of the steepest ascent while we are looking to go in the direction of the steepest descent. This is why there is a negative in the equation.
+#
+#
+# We will choose <code> step_size = 0.1 </code>:
+
+# +
+x = np.linspace(-5, 5, 100)
+plt.plot(x, y(x))  # plot y = x ** 2
+
+# We chose step size of 0.1
+step_size = 0.1
+
+# Define a function to get slope of y = x ** 2 at any x value:
+def dy_dx(x_input):
+    gradient = 2 * x_input  # derivative of y = x ** 2
+    return gradient
+
```

Comment:
Would it be possible to use PyTorch directly to calculate the gradients here? Either instead of the analytical gradient, or alongside it? I think that might make the code a bit shorter and underscore the point of why we're using autodifferentiation for optimization.